### ISSUE TITLE
Chaining multiple stages and persisting them using MLEM

### ISSUE THREAD

**Issue by drorata** [iterative/mlem#283]
(This is a follow up of a call with @aguschin)

In general, ML prediction consists of:

1. Features engineering
1. Features preprocessing
2. Model fitting

Once these steps are completed, predictions can be made. Note that when predicting value for an unseen data point, the features engineering and processing must take place prior to the actual prediction by the fitted model.

Alex outlined something like the following:

Feature engineering script that persists the function as a MLEM object.

```python
# feature-engineer.py
def add_features(df): 
    # do some features magic
    return df

mlem.api.save(add_features, “models/feature-processing”)
```

A training script which  

```python
# train.py
model = Model().fit(X,y)

mlem.api.save(model, “models/rf”)
```

For prediction purposes, something like the following can be used:

```python
# predict.py

fp = mlem.api.load(“models/feature-processing”)
model = mlem.api.load(“models/rf”)

model.predict(fp.transform(dataframe))
```

The question is now how to package this in a single structure. Doing something like:

```
$ mlem build docker models/rf 
```
 won't be aware of the feature engineering step.

Two options raised by Alex:

## Chanining:

```
$ mlem build docker models/feature-processing models/rf
```

## Creating a dedicated function that will be persisted:

```python
# prepare-model-for-deploy.py

fp = mlem.api.load(“models/feature-processing”)
model = mlem.api.load(“models/rf”)

def chain():
  return model.predict(fp.transform(dataframe))

mlem.api.save(chain, “models/chain”)
```

However, the second option is in many ways overlaps with `dvc.yaml`; at the end, the DAG defined by DVC is already targeting this need.

**Comment by drorata** [comment]
@dberenbaum @mike0sv FYI

**Comment by mike0sv** [comment]
The second option has an advantage because it should work right now. 
Though another disadvantage of it is that you will have an extra set of artifacts with basically the same contents.

I'm thinking we should have some other type of mlem object, which can have links to multiple mlem models inside. And this object can be used to configure those chains and then used as a substitute for model objects in methods like `apply/build/serve`

**Comment by drorata** [comment]
I another disadvantage of the 2nd option is that it introduces additional code whose only purpose is the integration with MLEM. It is not really needed for the ML process itself.

If all the features engineering and processing can be defined in terms of a scikit pipeline, then the resulting fitted pipeline can be persisted directly using MLEM. The problem is that in many cases (either due to complexity of the transformations or due to desire to build a modular code base or maybe due to other considerations) it is not possible to use scikit's pipelines.

I wonder how the DAG defined by `dvc.yaml` can be harnessed towards tackling this issue. See this diagram where I try to illustrate my intuition :)

<img src="https://user-images.githubusercontent.com/2905978/171870050-35f37f03-68ab-4fcd-9828-19b6122ad698.jpeg" width="400">

**Comment by mike0sv** [comment]
DVC pipeline will persist the data between stages, so if you have feature engineering and pre-processing as different stages you will have output of them both as files. And for inference you probably want to do all of it in-memory.
And also you cant re-use DAG for new data - you will have to create a new one, even if it is similar. And still it makes sense only if your test data is fixed (at least, not online)

**Comment by drorata** [comment]
A DVC pipeline is doing actually **two** things: (i) it defines the process and (ii) persists the data. The definition part (sometimes it is a subset of process; e.g. skipping data fetching for training) is the one that needs to be captured by MLEM. 

Maybe it is possible to "tag" stages in the process that need to be captured and persisted by MLEM?

**Comment by aguschin** [comment]
Also, it would be beneficial to have a Tutorial explaining how to work in this cases. Maybe we can apply DDD here to drive the development.

**Comment by aguschin** [comment]
A discussion on this: https://discord.com/channels/485586884165107732/903647230655881226/1016264255491813376

**Comment by dberenbaum** [comment]
I think you will typically need to define separate training and inference pipelines. Training is better kept modular, persisting to disk frequently to cache/debug stages, while for inference everything usually needs to happen in memory. Also, the inference pipeline depends on the training pipeline, not only for the model weights, but also for feature engineering things like defining the vocabulary in NLP models.

How can we make it easier to transition the training pipeline to inference? If you have a training pipeline like feature engineering -> preprocessing -> model fitting, you probably need a parallel inference/prediction stage that:
* Depends on raw data but also depends on the MLEM outputs from each training stage.
* Combines all those MLEM artifacts into one in-memory stage and can create a single MLEM artifact that builds/deploys those together.

Seems like the inference stage can include the custom/meta/chained function approach above as a first step. Once we have a basic way to at least define an inference pipeline in MLEM, we can start thinking about whether there are ways to further integrate with the existing DVC training pipeline to make it easier.

**Comment by aguschin** [comment]
Now you can add pre- and post-processors for your models. `mlem.api.save(..., preprocess=..., postprocess=...)` will apply `preprocess` before the model and `postprocess` after the model like this: `postprocess(model(preprocess(x)))`.

Added in #485, released in MLEM 0.4.0.

We're going to publish a blog post on this in Jan and add this to docs.
